File: 40. Combination Sum II.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int> > res;
        sort(candidates.begin(), candidates.end());
        vector<int> path;
        helper(candidates, 0, target, path, res);
        return res;
    }

    void helper(vector<int> &candidates, int idx, int target, vector<int> path, vector<vector<int> > &res) {
        if (target < 0) {
            return;
        } else if (target == 0) {
            res.push_back(path);
            return;
        } else if (idx >= candidates.size() || candidates[idx] > target) {
            return;
        }
        for (int i = idx; i < candidates.size(); i++) {
            if (i > idx && candidates[i] == candidates[i - 1]) continue;
            if (candidates[i] > target) break;
            path.push_back(candidates[i]);
            helper(candidates, i + 1, target - candidates[i], path, res);
            path.pop_back();
        }
    }
};

File: java/40. Combination Sum II.java
// O(2 ^ n)
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> res = new ArrayList<>();
        dfs(candidates, 0, target, new ArrayList<Integer>(), res);
        return res;
    }

    private void dfs(int[] candidates, int start, int target, ArrayList<Integer> path, List<List<Integer>> res) {
        if (target < 0) {
            return;
        }
        if (target == 0) {
            res.add(new ArrayList<>(path));
            return;
        }
        for (int i = start; i < candidates.length; ++i) {
            if (i == start || candidates[i] != candidates[i - 1]) {
                path.add(candidates[i]);
                dfs(candidates, i + 1, target - candidates[i], path, res);
                path.remove(path.size() - 1);
            }
        }
    }
}
